### ISSUE TITLE
Enhancement: [switch-exhaustiveness-check] considerDefaultExhaustiveForUnions should default to true

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#10307]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/switch-exhaustiveness-check/

### Description

The `considerDefaultExhaustiveForUnions` option was implemented to the `switch-exhaustiveness-check` rule 2 weeks ago by @developer-bandi in #9954. In the PR, the option was defaulted to false. However, I propose that the option should instead default to true.

When the option is turned off, it is only useful when *every switch statement* in a codebase has a block for *every single constituent*. However, I would argue that this is the exception rather than the norm. Consider a switch statement like this:

```ts
switch (roomType) {
  case RoomType.TREASURE: {
    return dataTreasureRooms;
  }

  case RoomType.BOSS: {
    return dataBossRooms;
  }

  case RoomType.SHOP: {
    return dataShopeRooms;
  }

  default: {
    return undefined;
  }
}
```

- This is a switch statement that is selecting specific values based on the value of a variable containing an enum.
- There are 30 different room types.
- Refactoring this switch statement to satisfy the lint rule would make it look like this:

```ts
switch (roomType) {
  case RoomType.TREASURE: {
    return dataTreasureRooms;
  }

  case RoomType.BOSS: {
    return dataBossRooms;
  }

  case RoomType.SHOP: {
    return dataShopeRooms;
  }

  case RoomType.DEFAULT:
  case RoomType.ERROR:
  case RoomType.MINI_BOSS:
  case RoomType.SECRET:
  case RoomType.SUPER_SECRET:
  case RoomType.ARCADE:
  case RoomType.CURSE:
  case RoomType.CHALLENGE:
  case RoomType.LIBRARY:
  case RoomType.SACRIFICE:
  case RoomType.DEVIL:
  case RoomType.ANGEL:
  case RoomType.DUNGEON:
  case RoomType.BOSS_RUSH:
  case RoomType.CLEAN_BEDROOM:
  case RoomType.DIRTY_BEDROOM:
  case RoomType.VAULT:
  case RoomType.DICE:
  case RoomType.BLACK_MARKET:
  case RoomType.GREED_EXIT:
  case RoomType.PLANETARIUM:
  case RoomType.TELEPORTER:
  case RoomType.TELEPORTER_EXIT:
  case RoomType.SECRET_EXIT:
  case RoomType.BLUE:
  case RoomType.ULTRA_SECRET: {
    return undefined;
  }
}
```

This code seems arguably worse. The whole point of the `switch-exhaustiveness-check` rule is that you get nice alerts in your codebase when new enum members are added. But in this case, we don't want to be alerted, because the default behavior is already handled. In other words, I only want to be alerted on the switch statements that do not have a `default` case.

At a high level, the question we should be asking is:
Is it more common to have switch statements that describe behavior for every single enum constituent?
Or is it more common to have switch statements that describe behavior for a subset of enum constituents?

I think the latter is true. If you agree, then it follows that the correct default value for the `considerDefaultExhaustiveForUnions` should be true.

**Comment by Zamiell** [comment]
> If a developers states that the cases which are explicitly tested are under control, he should rather go with a if / else if (= cases which are tested) - else (default for any other case) structure.

I completely disagree that switch statements should be replaced with `if else`. From what I have read, it is even considered an antipattern because it harms readability. See [this explanation](https://softwareengineering.stackexchange.com/a/15843). In short, using a if statement makes you have to read every single condition in the entire chain to make sure there are no surprised. However, in a switch statement, you immediately know the structure and can warp directly to where you want to go. I'm pretty sure I have also read blogs about this, but can't find them right now, maybe someone else can pitch in.

**Comment by OlivierZal** [comment]
I think you both agree, I'm just speaking about the purpose of the `switch-exhaustiveness-check` _by default_, which is about exhaustiveness. I don't promote `if-else` over `switch` though (even if we could challenge its usage when only few cases are handled over many others).

**Comment by kirkwaiblinger** [comment]
During review, I had it in my head that we were intentionally going with the breaking default from reading https://github.com/typescript-eslint/typescript-eslint/pull/9954#discussion_r1781091655, which I remember thinking was surprising, but I never really followed up on. Now that I'm looking more closely, I don't really see this explicitly discussed in that comment or elsewhere, and I'm wondering whether this was just a mistaken thought got stuck in my head. 

Note that there was a lot of confusion over the naming in general and we actually even released it with the opposite behavior from the name at first! See #10222

@JoshuaKGoldberg 

Also, FYI, the new option `considerDefaultExhaustiveForUnions' _is_ `false` by default. We made it so that the previous behavior now requires a specific opt-out by setting it to `true`.

**Comment by Zamiell** [comment]
> Also, FYI, the new option considerDefaultExhaustiveForUnions is false by default. We made it so that the previous behavior now requires a specific opt-out by setting it to true.

Sorry, that's a typo on my end. I'll update the OP accordingly.

**Comment by bradzacher** [comment]
From my anecdotal experience the majority of enums are small  -- much less than 10 members. Which means that having this option on by default makes sense because it's ergonomic to enumerate the options and it adds a lot of safety to do so in the codebase.

Massive enums with 30 members are definitely an outlier from my experience and are thus good candidates for a disable comment.

If such massive enums are the norm, rather than the exception in a codebase then it would likely be correct for that codebase to change the option.

To summarise: I think that the default is correct for the majority of codebases.

**Comment by kirkwaiblinger** [comment]
I do agree that it's the *right* default. So, whether or not it was fully intentional to release it as a breaking change... realistically either way I don't think we'll be changing it again now that it's out there. 

Closing this issue, but valid discussion and good reference issue to have in case people run into this unexpectedly when upgrading. Thanks for filing!

**Comment by Zamiell** [comment]
> From my anecdotal experience the majority of enums are small -- much less than 10 members.

Let's assume that the average number of enum members in a codebase is 5. Then, we could consider two versions of the same code. One using a `default` clause:

```ts
let numDoors = 4;

switch (roomType) {
  case RoomType.TREASURE: {
    numDoors += 1;
  }

  case RoomType.BOSS: {
    numDoors += 2;
  }

  default: {
    // Only in the above special cases should we increase the number of doors.
    break;
  }
}
```

And one without:

```ts
let numDoors = 4;

switch (roomType) {
  case RoomType.TREASURE: {
    numDoors += 1;
  }

  case RoomType.BOSS: {
    numDoors += 2;
  }
  
  case RoomType.ERROR:
  case RoomType.MINI_BOSS:
  case RoomType.SECRET: {
    // Only in the above special cases should we increase the number of doors.
    break;
  }
}
```

With `considerDefaultExhaustiveForUnions` set to false, consider what would happen when someone adds a new room to the `RoomType` enum:

- In the first code snippet, no errors/warnings would fire. This makes sense, because adding new enum members should not change the default behavior of the switch statement. We've explicitly established default behavior with the `default` keyword!
- In the second code snippet, an error/warning would fire, forcing us to go back to the code and ask the question: does this new room type need to have special behavior attached to it? Does this new room type have extra doors like Treasure Rooms do?

The punchline here is that **both forms of this code are valid and should be used for different purposes**:

- If the programmer knows that there is truly special behavior that should not apply to any new enum members in the future, then they should use a switch statement with a `default` case. By typing the word `default`, they explicitly indicate that future enum additions should fall under the `default` case, and that we should not be burdened in the future with busywork to superficially add new enum members to the block of code - that would be a waste of everybody's time.
- If the programmer knows that each enum member results in unique functionality, then should explicitly include every enum member in the switch statement, even if some of the enum members result in a no-op . By doing this, the programmer explicitly plants a flag on the code, knowing that the `switch-exhaustiveness-check` will force future programmers to come back to this exact place and ask the kind of questions like "should this new room type have extra doors like Treasure Rooms do?"

Brad's argument is that questions like "should this new room type have extra doors like Treasure Rooms do?" are good and safe, and that we should always force people to ask them. And that subsequently, the option should be defaulted to false, making large swathes of the ecosystem refactor the first form to the second.

But I feel like this is going too far. By default, having this option in will result in a ton of busywork in a codebase every time an enum member is added, **even when programmers have explicitly annotated with the `default` keyword that no future updates are necessary**. That will make programmers hate the `switch-exhaustiveness-check` rule. And they might turn it off entirely without realizing that this option is customizable.

Obviously, we want to strike the balance between creating a ton of noise for end-users, and keeping them safe. I feel that with this option, we are tipping too far into the "noise" side, and that it does not represent a good default.

**Comment by bradzacher** [comment]
From my experience at scale - this is the default you want. You want someone to consider if the new case was meant to be covered by the `default` or not.

Most logic s around enums I've seen have been because of this exact case - someone forgets to add a case to a switch with a default and it causes a crash. This is the reason that the flow team built a similar rule years ago and all enums in the gigantic FB codebase of >900k files have this very logic applied to it.

Safety and correctness is worth a few extra lines of verbosity.

It entirely depends on the scale you're working at. If you're working in a codebase that's small enough that you can keep most if it in your head at once then yeah this option is going to create noise for you. It would create noise in our codebase here for example.
But most codebases tend to be large enough that you can't keep it in one person's head so the check helps with catching forgotten cases.
